src/main/java/ua/bookstore/online/repository/book/BookSpecificationBuilder.java

```diff
@@ -0,0 +1,39 @@
+package ua.bookstore.online.repository.book;
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+import ua.bookstore.online.dto.BookSearchParameters;
+import ua.bookstore.online.model.Book;
+import ua.bookstore.online.repository.SpecificationBuilder;
+import ua.bookstore.online.repository.SpecificationProviderManager;
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParameters bookSearchParameters) {
+        Specification<Book> specification = Specification.where(null);
+        String[] titles = bookSearchParameters.getTitles();
+        if (titles != null && titles.length > 0) {
+            specification = specification.and(bookSpecificationProviderManager
+                    .getSpecificationProvider(BookSearchParameter.TITLE.getName())
+                    .getSpecification(titles));
+        }
+        String[] authors = bookSearchParameters.getAuthors();
+        if (authors != null && authors.length > 0) {
+            specification = specification.and(bookSpecificationProviderManager
+                    .getSpecificationProvider(BookSearchParameter.AUTHOR.getName())
+                    .getSpecification(authors));
+        }
+        String[] isbns = bookSearchParameters.getIsbns();
+        if (isbns != null && isbns.length > 0) {
+            specification = specification.and(bookSpecificationProviderManager
+                    .getSpecificationProvider(BookSearchParameter.ISBN.getName())
+                    .getSpecification(isbns));
+        }
```

Comment:
I see at least 2 code duplications:

1. Check if the collection is null or empty. I suggest to create some util class for such checks or use already existing libraries like Apache common utils
2. Code that populates specification is common for both author and title, so I suggest extracting it in method and reuse. This will make code more flexible because it will be much easier to add new search parameters, maybe event strategy will be suitable here but it may be overcomplication, so just make a method for now

Reply:
It would be possible to make specification using loop